### ISSUE TITLE
MQTT Failed to Reconnect

### ISSUE THREAD

**Issue by pauldeng** [u-blox/ubxlib_cellular_applications_xplr_iot#20]
Hi repo owner,

I still have this issue even with firmware upgraded and using your latest code.

The MQTT failed to reconnect after network resumed.

Here is detail:
- Firmware version 3.30
- Your latest code on branch v1.0-DriveTestFixes
- Operate normal at the beginning
- Put it into Faraday cage
- last message with timestamp 1692933591
- Take it out
- first and only message after is timestamp 1692933621
- Received one MQTT message
- No more messages
- Waited 3 minutes, but no reconnect
- errValue is 34, which is not defined in enum

```log
...
03:19:51.129: {"Timestamp":1692933591,"CellQuality":{"RSRP":-88,"RSRQ":-7,"RSSI":-59,"SNR":9,"RxQual":5},"CellInfo":{"LogicalCellID":"0x08319d03","PhysicalCellID":"192","EARFCN":9410,"PLMN":50501,"Operator":"Telstra Mobile"}}
03:19:52.338: {"Timestamp":1692933591,"Location":{"Altitude":151987,"Latitude":-37.8366068,"Longitude": 145.1827036,"Accuracy":12663,"Speed":46,"GNSSTimestamp":1692933593}}
03:20:21.196: {"Timestamp":1692933621,"CellQuality":{"RSRP":-140,"RSRQ":-31,"RSSI":0,"SNR":-17,"RxQual":-1},"CellInfo":{"LogicalCellID":"0x08319d03","PhysicalCellID":"192","EARFCN":9410,"PLMN":50501,"Operator":"Telstra Mobile"}}
03:20:22.406: {"Timestamp":1692933621,"Location":{"Altitude":158484,"Latitude":-37.8,"Longitude": 145.1,"Accuracy":18284,"Speed":22,"GNSSTimestamp":1692933623}}
03:20:51.263: {"Timestamp":1692933651,"CellQuality":{"RSRP":-140,"RSRQ":-31,"RSSI":0,"SNR":-17,"RxQual":-1},"CellInfo":{"LogicalCellID":"0x08319d03","PhysicalCellID":"192","EARFCN":9410,"PLMN":50501,"Operator":"Telstra Mobile"}}
03:21:11.419: Network status unknown.
03:21:11.423: Network status unknown.
03:21:25.937: Network connection is down...
03:21:56.124: Network connection is down...
03:22:26.474: Network connection is down...
03:22:56.663: Network connection is down...
03:23:27.012: Network connection is down...
03:23:57.201: Network connection is down...
03:24:27.557: Network connection is down...
03:24:57.908: Network connection is down...
03:25:28.095: Network connection is down...
03:25:58.451: Network connection is down...
03:26:28.638: Network connection is down...
03:26:58.994: Network connection is down...
03:27:29.183: Network connection is down...
03:27:59.545: Network connection is down...
03:28:29.732: Network connection is down...
03:29:00.093: Network connection is down...
03:29:30.283: Network connection is down...
03:30:00.649: Network connection is down...
03:30:30.837: Network connection is down...
03:31:01.202: Network connection is down...
03:31:31.390: Network connection is down...
03:32:01.758: Network connection is down...
03:32:31.946: Network connection is down...
03:33:02.315: Network connection is down...
03:33:32.502: Network connection is down...
03:34:02.872: Network connection is down...
03:34:33.061: Network connection is down...
03:35:03.434: Network connection is down...
03:35:33.622: Network connection is down...
03:36:03.997: Network connection is down...
03:36:34.186: Network connection is down...
03:37:04.568: Network connection is down...
03:37:34.756: Network connection is down...
03:38:05.134: Network connection is down...
03:38:35.322: Network connection is down...
03:39:05.700: Network connection is down...
03:39:15.693: Network is Registered: Roaming
03:39:24.417: {"Timestamp":1692934764,"Location":{"Altitude":152920,"Latitude":-37.8,"Longitude": 145.1,"Accuracy":12693,"Speed":58,"GNSSTimestamp":1692934766}}
03:39:24.436: {"Timestamp":1692934764,"CellQuality":{"RSRP":-80,"RSRQ":-8,"RSSI":-51,"SNR":7,"RxQual":5},"CellInfo":{"LogicalCellID":"0x08319d03","PhysicalCellID":"192","EARFCN":9410,"PLMN":50501,"Operator":"Telstra Mobile"}}

*** WARNING ************************************************
03:39:28.325: Failed to publish MQTT message,  error: 34


*** WARNING ************************************************
03:39:32.300: Failed to publish MQTT message,  error: 34

03:39:54.026: {"Timestamp":1692934794,"CellQuality":{"RSRP":-84,"RSRQ":-10,"RSSI":-57,"SNR":13,"RxQual":5},"CellInfo":{"LogicalCellID":"0x08319d03","PhysicalCellID":"192","EARFCN":9410,"PLMN":50501,"Operator":"Telstra Mobile"}}
03:39:57.839: {"Timestamp":1692934794,"Location":{"Altitude":149593,"Latitude":-37.8,"Longitude": 145.1,"Accuracy":12515,"Speed":40,"GNSSTimestamp":1692934796}}

*** WARNING ************************************************
03:39:57.906: Failed to publish MQTT message,  error: 34


*** WARNING ************************************************
03:40:01.796: Failed to publish MQTT message,  error: 34

03:40:24.099: {"Timestamp":1692934824,"CellQuality":{"RSRP":-86,"RSRQ":-7,"RSSI":-59,"SNR":12,"RxQual":5},"CellInfo":{"LogicalCellID":"0x08319d03","PhysicalCellID":"192","EARFCN":9410,"PLMN":50501,"Operator":"Telstra Mobile"}}
03:40:27.911: {"Timestamp":1692934824,"Location":{"Altitude":141718,"Latitude":-37.8,"Longitude": 145.1,"Accuracy":12674,"Speed":44,"GNSSTimestamp":1692934826}}

*** WARNING ************************************************
03:40:28.066: Failed to publish MQTT message,  error: 34


*** WARNING ************************************************
03:40:31.946: Failed to publish MQTT message,  error: 34

03:40:54.165: {"Timestamp":1692934854,"CellQuality":{"RSRP":-86,"RSRQ":-9,"RSSI":-57,"SNR":10,"RxQual":5},"CellInfo":{"LogicalCellID":"0x08319d03","PhysicalCellID":"192","EARFCN":9410,"PLMN":50501,"Operator":"Telstra Mobile"}}
03:40:57.978: {"Timestamp":1692934854,"Location":{"Altitude":136204,"Latitude":-37.8365944,"Longitude": 145.1826673,"Accuracy":11966,"Speed":28,"GNSSTimestamp":1692934856}}

*** WARNING ************************************************
03:40:58.045: Failed to publish MQTT message,  error: 34


*** WARNING ************************************************
03:41:01.935: Failed to publish MQTT message,  error: 34

03:41:24.232: {"Timestamp":1692934884,"CellQuality":{"RSRP":-85,"RSRQ":-15,"RSSI":-57,"SNR":9,"RxQual":5},"CellInfo":{"LogicalCellID":"0x08319d03","PhysicalCellID":"192","EARFCN":9410,"PLMN":50501,"Operator":"Telstra Mobile"}}
03:41:28.045: {"Timestamp":1692934884,"Location":{"Altitude":134022,"Latitude":-37.8365804,"Longitude": 145.1826436,"Accuracy":12962,"Speed":44,"GNSSTimestamp":1692934886}}

*** WARNING ************************************************
03:41:28.230: Failed to publish MQTT message,  error: 34


*** WARNING ************************************************
03:41:32.110: Failed to publish MQTT message,  error: 34

03:41:54.300: {"Timestamp":1692934914,"CellQuality":{"RSRP":-89,"RSRQ":-7,"RSSI":-63,"SNR":14,"RxQual":5},"CellInfo":{"LogicalCellID":"0x08319d03","PhysicalCellID":"192","EARFCN":9410,"PLMN":50501,"Operator":"Telstra Mobile"}}
03:41:58.112: {"Timestamp":1692934914,"Location":{"Altitude":147857,"Latitude":-37.8366184,"Longitude": 145.1827523,"Accuracy":13054,"Speed":453,"GNSSTimestamp":1692934916}}

*** WARNING ************************************************
03:41:58.179: Failed to publish MQTT message,  error: 34


*** WARNING ************************************************
03:42:02.069: Failed to publish MQTT message,  error: 34

03:42:24.366: {"Timestamp":1692934944,"Location":{"Altitude":149014,"Latitude":-37.8365420,"Longitude": 145.1828051,"Accuracy":27870,"Speed":963,"GNSSTimestamp":1692934946}}
03:42:24.385: {"Timestamp":1692934944,"CellQuality":{"RSRP":-96,"RSRQ":-7,"RSSI":-63,"SNR":11,"RxQual":4},"CellInfo":{"LogicalCellID":"0x08319d03","PhysicalCellID":"192","EARFCN":9410,"PLMN":50501,"Operator":"Telstra Mobile"}}

*** WARNING ************************************************
03:42:28.263: Failed to publish MQTT message,  error: 34


*** WARNING ************************************************
03:42:32.232: Failed to publish MQTT message,  error: 34

03:42:54.450: {"Timestamp":1692934974,"Location":{"Altitude":152739,"Latitude":-37.8367386,"Longitude": 145.1836175,"Accuracy":41223,"Speed":4831,"GNSSTimestamp":1692934976}}
03:42:54.469: {"Timestamp":1692934974,"CellQuality":{"RSRP":-97,"RSRQ":-9,"RSSI":-63,"SNR":7,"RxQual":3},"CellInfo":{"LogicalCellID":"0x08319d03","PhysicalCellID":"192","EARFCN":9410,"PLMN":50501,"Operator":"Telstra Mobile"}}

*** WARNING ************************************************
03:42:58.357: Failed to publish MQTT message,  error: 34


*** WARNING ************************************************
03:43:02.233: Failed to publish MQTT message,  error: 34

03:43:17.469: Exit button pressed, closing down... Please wait for the RED LED to go out...
03:43:23.904: Waiting for app tasks to stop... This can take sometime if waiting for AT commands to timeout...
03:43:24.016: All tasks have now finished...
03:43:24.021: Stop Registration task requested...
03:43:24.027: Waiting for Registration task to stop...
03:43:24.047: De-registering from the network...
03:43:24.835: Deregistered from cellular network
03:43:26.032: Application Finished.

********************************************************
*** LOG END ********************************************
********************************************************
```

Cheers,
Paul

**Comment by pauldeng** [comment]
Unfortunately, I do not have the full log.
Microwave oven is my cage and could not have usb cable out.
Need to leave the device in for lets say 20 minutes or more to reproduce it.

**Comment by philwareublox** [comment]
You can press button #1 when your turn on the device in the first 5 seconds when the blue LED is shown to display the log again.

Turn on the XPLR and RED LED shows: This gives you a few seconds to "connect" to the COM Port
BLUE LED shows for 5 seconds: This is waiting for either button #1 or button #2 to be pressed.
Button #1 - display log file
Button #2 - delete log file

I will have a look to see if I can reproduce this.

**Comment by philwareublox** [comment]
What MQTT are you using? MQTT or MQTT-SN?

MQTT Error Code: 34 - No Network Service
MQTT-SN Error Code: 34 - Incoming message cannot be saved, the buffer is full

From the error code which makes more sense I assume you are using MQTT and not MQTT-SN

**Comment by pauldeng** [comment]
I am using my own rabbitmq mqtt broker, not using mqtt-sn.

Here is [full log](https://ax534dxu526w.compat.objectstorage.ap-melbourne-1.oraclecloud.com/tmp/u-blox-xplr-iot.log?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=2854cbd40f7bcc7e1287de30dc97c08f0daa3e33%2F20230825%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20230825T094237Z&X-Amz-Expires=604800&X-Amz-SignedHeaders=host&X-Amz-Signature=2ef01f3796def46e5d7f124f772ab33b44b18e2d9945a69af6d9ca985976ed68), contains multiple tests.

**Comment by philwareublox** [comment]
Thanks for the log - there are some interesting bits in this, even when the network hasn't gone down you still have the MQTT client saying the there is no network service... 
Do you know what firmware is running in the SARA-R5? What firmware did you update to?
(I will add some logs on the cellular module and commit it to the v1.0 update branch in a moment)

**Comment by pauldeng** [comment]
This error 34 is different from turn off mqtt broker intentionally.

If mqtt broker is turned off, it is keep trying:
```log
09:50:06.073: MQTT client disconnected, trying to connect...
09:50:06.080: Connecting to MQTT Broker on abc.example.domain:port_number...

************************************************************
*** ERROR **************************************************
09:50:13.565: Failed to connect to the MQTT Broker: -9

09:50:18.587: MQTT client disconnected, trying to connect...
09:50:18.594: Connecting to MQTT Broker on abc.example.domain:port_number...
```

**Comment by pauldeng** [comment]
Here is my device info
![image](https://github.com/u-blox/ubxlib_cellular_applications_xplr_iot/assets/341848/6b81edc0-72db-4966-834a-aad2345f3ac2)
upgraded from this [firmware](https://content.u-blox.com/sites/default/files/2023-06/SARA-R510S-01B-01-IP-0330A0001-005K00.zip?_ga=2.161557564.1643030297.1692868092-1495629776.1692868022).

**Comment by philwareublox** [comment]
Yes, if you turn off the broker it is very different to the 34 error, as this is an internal network status it is reading.

You are using the latest firmware. 

I will try and see if I can reproduce this.... but really there is nothing the XPLR application can do, as far as it knows, the network is down, but obviously it is not...

**Comment by pauldeng** [comment]
If we can treat error 34 as error 9, but it could be ubxlib thing as you mentioned.

Foil wraps it, leave it inside of upside down microwave, wait 20 minutes and hopefully you can reproduce this.

**Comment by philwareublox** [comment]
It's nothing to do with ubxlib either, this is a sara-r5 module issue. ubxlibx just reports the status and the xplr application uses those statuses to know if to send a message or not.

**Comment by philwareublox** [comment]
@pauldeng, Could you recompile the XPLR application with the debug level set to debug and also turn on the ubxlib logging and post the log file here.

Set to eDEBUG:
https://github.com/u-blox/ubxlib_cellular_applications_xplr_iot/blob/b5fd50f2fc6e71d62f93363aeb94a1565628bf2e/applications/cellular_tracker/config/config.h#L52

Delete the // so that the #define is enabled.
https://github.com/u-blox/ubxlib_cellular_applications_xplr_iot/blob/b5fd50f2fc6e71d62f93363aeb94a1565628bf2e/applications/common/appInit.c#L37

(I should move the UBXLIB logging #define to the config.h file really...)

thanks.

**Comment by pauldeng** [comment]
What is your email address?
My log contains sensitive info.

**Comment by pauldeng** [comment]
Sent to your email phil.ware@u-blox.com <phil.ware@u-blox.com>

**Comment by philwareublox** [comment]
This issue is because of a bug in the R510. When the module goes out-of-coverage for an extended period the socket used by the internal MQTT client is broken and the MQTT client just keeps trying to use it. It returns error 34 for each publish message. The work-around is to disconnect and reconnect to the broker. 
This has been implemented in the latest branch v1.0-drive tests.
